File: frontend/src/lib/quiz/modules/multiplication.ts
import { BaseQuizModule, Question } from "../base_module";

export class MultiplicationModule extends BaseQuizModule {
  slug = "multiplication";
  title = "Multiplication Tables";
  description = "Master your times tables with configurable ranges.";

  generateQuestions(config: Record<string, any>): Question[] {
    const rangeStart = Number(config.range_start ?? 2);
    const rangeEnd = Number(config.range_end ?? 12);
    const length = Number(config.length ?? 10);
    
    const easyNums = new Set([0, 1, 10, 11]);
    const questions: Question[] = [];
    
    for (let tableNum = rangeStart; tableNum <= rangeEnd; tableNum++) {
      if (easyNums.has(tableNum)) continue;
      for (let multiplier = 1; multiplier <= length; multiplier++) {
        if (easyNums.has(multiplier)) continue;
        questions.push({
          question_string: `${tableNum} x ${multiplier}`,
          correct_answer: tableNum * multiplier,
        });
      }
    }
    
    // Shuffle
    for (let i = questions.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [questions[i], questions[j]] = [questions[j], questions[i]];
    }
    
    return questions;
  }

  evaluateAnswer(question: Question, userAnswer: any): boolean {
    const parsed = parseInt(String(userAnswer), 10);
    if (isNaN(parsed)) return false;
    return parsed === question.correct_answer;
  }
}

File: backend/app/services/modules/multiplication.py
import random
from typing import Dict, Any, List
from .base_module import BaseQuizModule, Question

class MultiplicationQuestion(Question):
    def __init__(self, num1: int, num2: int):
        self.question_string = f"{num1} x {num2}"
        self.correct_answer = num1 * num2

class MultiplicationModule(BaseQuizModule):
    slug = "multiplication"
    title = "Multiplication Tables"
    description = "Master your times tables with configurable ranges."

    def generate_questions(self, config: Dict[str, Any]) -> List[Question]:
        '''
        Expects config:
        - range_start: int (e.g. 2)
        - range_end: int (e.g. 12)
        - length: int (e.g. 10 to test up to x10)
        '''
        range_start = config.get("range_start", 2)
        range_end = config.get("range_end", 12)
        length = config.get("length", 10)
        
        easy_nums = {0, 1, 10, 11}
        questions = []
        for table_num in range(range_start, range_end + 1):
            if table_num in easy_nums:
                continue
            for multiplier in range(1, length + 1):
                if multiplier in easy_nums:
                    continue
                questions.append(MultiplicationQuestion(table_num, multiplier))
                
        # Shuffle internally to gamify it
        random.shuffle(questions)
        return questions

    def evaluate_answer(self, question: Question, user_answer: Any) -> bool:
        try:
            return int(user_answer) == question.correct_answer
        except ValueError:
            return False

# Registering automatically allows plugin-style architecture
def register() -> BaseQuizModule:
    return MultiplicationModule()
